rfc-176-replacing-static.md

```diff
@@ -0,0 +1,127 @@
+
+### Static Assets (move to `govuk_web_layout`)
+
+#### How GOV.UK handles assets
+
+GOV.UK uses the standard rails assets pipeline for creation of assets, in which at asset compile time a fingerprint is generated for each asset based on the asset’s content, and during HTML page rendering the base names of the assets are replaced with the fingerprinted version. This allows clients to implement a pretty simple caching policy - they can always cache a given asset for as long as they like, because the asset’s name will change if the content changes - if the client doesn’t already have that fingerprinted file, they will need to download it.
+
+Asset pre-compilation happens when the container image is built (during the ‘deploy’ GitHub Actions workflow). This process generates the compiled assets and the Sprockets manifest. The manifest is the mechanism for Rails to map the original asset name (`crown.png`) to the fingerprinted asset name (`crown-b2ba232ade53223.png`), so the deployed app has a record of everything that was compiled as part of its assets:precompile step.
+
+Upon deployment into the Kubernetes cluster, a job runs which copies anything from `/app/public/assets/<name of app>` into an S3 bucket `s3://govuk-app-assets-<environment>/assets/<name of app>`. [Fastly maps] any call for (e.g.) `https://www.gov.uk/assets/` to that bucket, so a request for `https://www.gov.uk/assets/frontend/crown-b2ba232ade53223.png` will be served from that s3 bucket.
+
+
+![Current asset deploy pathway](rfc-176/current_govuk_asset_pathway.png)
+
+#### What this means for shared assets
+Because the upload-assets job in [govuk-helm-charts] includes the app name, there currently isn’t a mechanism for apps to upload assets in a shared context. If two apps both precompile a CSS file for the Table component, for instance, the fingerprint will be the same, but the path will be different, so a client that views two pages with Tables on (one in [frontend], one in [government-frontend]) will download the same file twice.
+
+However, because all of these apps upload to different folders in the same S3 bucket, it would be relatively easy to add a second step to the upload which took assets from (say) `/app/public/assets/shared` on the deployed images and upload that to a shared folder. Because the same asset compiled in two different apps ends up with the same name, the second app’s uploaded asset would overwrite the first one with the exact same name and data. As long as mechanisms were in place to prevent confusion from two apps uploading at the same time (this may already be present in S3), we could safely write to the same shared assets directory.
```

Comment:
Do we need to treat these as shared assets? What if we were to keep these assets as if they were specific to individual application? Obviously, you'd have duplication of what are essentially the same asset, but less complexity and minimal change the existing asset upload job?

Reply:
Shared assets are desirable when it comes to css and js - and possibly some images. Otherwise the user downloads them multiple times for no gain during a user journey.

Reply:
Think it'd be good to understand the realistic user impact vs engineering effort/possible risks of using shared assets. Assets will likely already be cached at the CDN level, then cached in browser for their lifetime (therefore impact is amortised across multiple user journeys). This will also start to improve once we start to consolidate frontend apps. Can these assets from the gems be bundled into existing assets? Whats the worst case - i.e. how many extra requests/data transferred for say a user journey across 4-6 frontend apps on cache cleared browser?

Reply:
We're encouraged in the frontend discipline to consider the desperate user on lowest spec phone on the shonkiest connection - we might only be talking about a few kilobytes here and there, and only on their first user journey, but regardless of it being probably low impact for most in the grand scheme of things, it seems needlessly unnecessary to be delivering the same content multiple times.

Reply:
So: @theseanything I'd be interested to hear what the worries are around complexity of the upload job, because I considered this an okay plan on the basis that any changes would be minimal at most. In the diagram I've made this two jobs, but that's entirely unnecessary - we should be able to do this in the existing job with only a few lines changed - I've knocked up a draf PR here that should do it? - https://github.com/alphagov/govuk-helm-charts/pull/2265

Reply:
Okay, sounds fine if you're confident the complexity and effort are low. The complexity around serving a subset of assets from a different path could be point of confusion. But primarily, was worried about application being dependent on assets produced by other application - but I guess we're saved a bit by the fingerprinting hash in the filename. (I.e. if another rendering application renders the templates differently. e.g. different versions of the gems on each frontend app. They'll produce a different asset name, but you'll end up with multiple requests then anyway).

Reply:
Yeah, I think absent a major change in Sprockets we're safe there, as the fingerprint is determined solely by the contents of the file, so different apps would always create the same fingerprinted name for an asset if the asset contents are identical.

Reply:
As per comments below, I'm going to simplify this somewhat - we actually don't need shared and app-only components to go in separate folders. A single assets folder can have a mix of shared components and components from disparate apps, and the fingerprinting will keep them all separate. The bucket doesn't need to have any organisation.

Reply:
(Note - I'm going to mark this as resolved as it's moved out of the scope of the RFC. We will look at a simpler way to share assets, but it's not part of the path to removing Static)